module/zfs/dbuf.c

```diff
@@ -3074,12 +3131,25 @@ dbuf_write_ready(zio_t *zio, arc_buf_t *buf, void *vdb)
 		}
 	} else {
 		blkptr_t *ibp = db->db.db_data;
+		int epbs = dn->dn_phys->dn_indblkshift - SPA_BLKPTRSHIFT;
 		ASSERT3U(db->db.db_size, ==, 1<<dn->dn_phys->dn_indblkshift);
 		for (i = db->db.db_size >> SPA_BLKPTRSHIFT; i > 0; i--, ibp++) {
 			if (BP_IS_HOLE(ibp))
 				continue;
 			fill += BP_GET_FILL(ibp);
 		}
+		/*
+		 * it turns out that pre-filling Ln holes in dbuf_read_impl()
+		 * can result in metadata block ids pointing beyond the
+		 * dn_maxblkid; this seems like a good place to fix dn_maxblkid
+		 * up in a manner similar to one above for level-0 blocks
+		 */
+		mutex_enter(&dn->dn_mtx);
+		if (db->db_blkid >
+			dn->dn_phys->dn_maxblkid >> (db->db_level * epbs))
+			dn->dn_phys->dn_maxblkid = ((db->db_blkid + 1) <<
```

Comment:
I think this violates the intended meaning of dn_maxblkid; I think the idea is that if the file is empty, dn_maxblkid is 0, and otherwise it's the number of the highest blkid of an actual filled in block on that object.  With this new proposal, you could easily end up in situations where dn_maxblkid is much farther along than the last filled-in l0 block.